src/util/ieee_float.cpp

```diff
@@ -0,0 +1,167 @@
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "log.h"
+#include "porting.h"
+#include <stdint.h>
+#include <math.h>
+
+// Given an unsigned 32-bit integer representing an IEEE-754 single-precision
+// float, return the float.
+float u32_FloatSlow(uint32_t i)
+{
+	float res;
+	int exp = (i >> 23) & 0xFF;
+	uint32_t sign = i & 0x80000000UL;
+	uint32_t imant = i & 0x7FFFFFUL;
+	if (exp == 0xFF) {
+		// Inf/NaN
+		if (imant == 0)
+			return sign ? -1.f/0.f : 1.f/0.f;
+		return 0.f/0.f; // always return indeterminate
+	}
+
+	if (!exp) {
+		// Denormal or zero
+		return sign ? -ldexpf((float)imant, -149) : ldexpf((float)imant, -149);
+	}
+
+	return sign ? -ldexpf((float)(imant | 0x800000), exp - 150) :
+		ldexpf((float)(imant | 0x800000), exp - 150);
+}
+
+// Given a float, return an unsigned 32-bit integer representing the float
+// in IEEE-754 single-precision format.
+uint32_t float_u32Slow(float f)
+{
+	uint32_t signbit = copysignf(1.0f, f) == 1.0f ? 0 : 0x80000000UL;
+	uint32_t imant;
+	int exp;
+	if (f == 0.f)
+		return signbit;
+	if (isnanf(f))
+		return signbit | 0x7FC00000UL;
+	if (isinff(f))
+		return signbit | 0x7F800000UL;
+	float mant = frexpf(f, &exp);
+	imant = (uint32_t)floorf((signbit ? -16777216.f : 16777216.f) * mant);
+	exp += 126;
+	if (exp <= 0) {
+		// Denormal
+		return signbit | (imant >> 1 - exp);
+	}
+
+	if (exp >= 255) {
+		// Overflow due to the platform having exponents bigger than IEEE ones.
+		// Return signed infinity.
+		return signbit | 0x7F800000UL;
+	}
+
+	// Regular number
+	return signbit | (exp << 23) | (imant & 0x7FFFFFUL);
+}
+
+
+// Test the functions.
+// This test needs the following requisites in order to work:
+// - The compiler must support uint32_t.
+// - The compiler must treat union by overlapping the memory of its fields.
+// - The float type must be a 32 bits IEEE-754 single-precision float.
+// - The endianness of floats and integers must match.
+// - 0.f/0.f must quietly return a NaN.
+int getFloatSerializationType()
+{
+	union {
+		float f;
+		uint32_t i;
+	} fui;
+
+	float f;
+	uint_fast32_t i;
+	char buf[128];
+
+	if (sizeof(fui) != 4) {
+		infostream << "Floats are not 32 bits in this platform" << std::endl;
+		return 1;
+	}
+
+	fui.i = 0;
+	fui.f = -22220490.f;
+	if (fui.i != 0xCBA98765UL) {
+		infostream << "Invalid endianness or machine not IEEE-754 compliant" << std::endl;
```

Comment:
This test in particular checks the endianness of float with respect to integer, and tests only for ABCD. It won't allow the DCBA etc. tests to be reached.